### PR TITLE
Optimize a bunch of distance comparisons

### PR DESCRIPTION
**Pull request by Tayrtahn**
Requires space-wizards/RobustToolbox#4999

## About the PR

Replaced a bunch of distance comparisons with more optimized versions.

## Why / Balance

Faster code.

## Technical details

A bunch of places in the code did range checks or other distance comparisons by calculating a vector length. Vector lengths is relatively expensive to calculate though, as they require finding a square root. If the actual length isn't needed (only a comparison to another value), it's cheaper to just compare the square distances instead.

This PR uses some new helper methods in RT to replace many instances of `if (someVector.Length() > someValue)` with `if (someVector.IsLongerThan(someValue))`, eliminating a square root calculation. In some instances, the actual length was needed later in the code so it couldn't be eliminated, so the square root calculation was instead delayed until after some conditional checks were passed, so that it won't be wasted when the method returns early.
i.e:
```C#
var length = someVector.Length();
if (length > someValue)
   return;
SomeMethodThatNeedsLength(length, foo);
```
becomes
```C#
if (someVector.IsLongerThan(someValue))
   return;
SomeMethodThatNeedsLength(someVector.Length(), foo);
```

## Media

Code
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

### CODE REVIEW

**Review by PJB3005 — COMMENTED** [review]
Honestly just looking at this diff I'd rather see a helper method made for this.

**Comment by Tayrtahn** [comment]
> Honestly just looking at this diff I'd rather see a helper method made for this.

That's a good call and now I'm annoyed that I didn't do it that way in the first place. Will change!

File: Content.Server/Fluids/EntitySystems/SpraySystem.cs [hunk 1]
```diff
@@ -111,7 +111,7 @@ private void OnAfterInteract(Entity<SprayComponent> entity, ref AfterInteractEve
                 break;
 
             // Spawn the vapor cloud onto the grid/map the user is present on. Offset the start position based on how far the target destination is.
-            var vaporPos = userMapPos.Offset(distance < 1 ? quarter : threeQuarters);
+            var vaporPos = userMapPos.Offset(diff.IsShorterThan(1) ? quarter : threeQuarters);
```
Comment (moonheart08) [inline Content.Server/Fluids/EntitySystems/SpraySystem.cs:114]: This does the length calculation twice (minus sqrt), which is slower than doing it once with the square root. I think you over-estimate how expensive square root is.

File: Content.Server/Pinpointer/PinpointerSystem.cs [hunk 2]
```diff
@@ -187,12 +187,11 @@ protected override void UpdateDirectionToTarget(EntityUid uid, PinpointerCompone
...
     {
-        var dist = vec.Length();
-        if (dist <= pinpointer.ReachedDistance)
+        if (vec.IsShorterThanOrEqualTo(pinpointer.ReachedDistance))
             return Distance.Reached;
-        else if (dist <= pinpointer.CloseDistance)
+        else if (vec.IsShorterThanOrEqualTo(pinpointer.CloseDistance))
             return Distance.Close;
-        else if (dist <= pinpointer.MediumDistance)
+        else if (vec.IsShorterThanOrEqualTo(pinpointer.MediumDistance))
```
Comment (moonheart08) [inline Content.Server/Pinpointer/PinpointerSystem.cs:194]: This is slower than what we had before. You overestimate how expensive sqrt is vs `x^2 + y^2`

For a rough approximation, the cost of the sqrt is that of the two squaring operations (not the addition). It's very minor and only adds up in hot loops.

File: Content.Server/Singularity/EntitySystems/SingularityAttractorSystem.cs [hunk 3]
```diff
@@ -78,11 +78,10 @@ private void Update(EntityUid uid, SingularityAttractorComponent? attractor = nu
                 continue;
 
             var biasBy = mapPos.Position - singuloMapPos.Position;
-            var length = biasBy.Length();
-            if (length <= MinAttractRange)
+            if (biasBy.IsLongerThanOrEqualTo(MinAttractRange))
                 return;
 
-            biasBy = Vector2.Normalize(biasBy) * (attractor.BaseRange / length);
+            biasBy = Vector2.Normalize(biasBy) * (attractor.BaseRange / biasBy.Length());
```
Comment (moonheart08) [inline Content.Server/Singularity/EntitySystems/SingularityAttractorSystem.cs:84]: Slower than before.

File: Content.Server/SurveillanceCamera/Systems/SurveillanceCameraMicrophoneSystem.cs [hunk 4]
```diff
@@ -31,20 +31,21 @@ private void OnExpandRecipients(ExpandICChatRecipientstEvent ev)
...
+            if (diff.IsShorterThan(0) || diff.IsLongerThan(ev.VoiceRange))
                 continue;
 
             foreach (var viewer in camera.ActiveViewers)
             {
                 // if the player has not already received the chat message, send it to them but don't log it to the chat
                 // window. This is simply so that it appears in camera.
                 if (TryComp(viewer, out ActorComponent? actor))
-                    ev.Recipients.TryAdd(actor.PlayerSession, new ICChatRecipientData(range, false, true));
+                    ev.Recipients.TryAdd(actor.PlayerSession, new ICChatRecipientData(diff.Length(), false, true));
```
Comment (moonheart08) [inline Content.Server/SurveillanceCamera/Systems/SurveillanceCameraMicrophoneSystem.cs:48]: Slower than before.

File: Content.Shared/Damage/Systems/DamageOnHighSpeedImpactSystem.cs [hunk 5]
```diff
@@ -47,7 +45,7 @@ private void HandleCollide(EntityUid uid, DamageOnHighSpeedImpactComponent compo
         if (_robustRandom.Prob(component.StunChance))
             _stun.TryStun(uid, TimeSpan.FromSeconds(component.StunSeconds), true);
 
-        var damageScale = component.SpeedDamageFactor * speed / component.MinimumSpeed;
+        var damageScale = component.SpeedDamageFactor * args.OurBody.LinearVelocity.Length() / component.MinimumSpeed;
```
Comment (moonheart08) [inline Content.Shared/Damage/Systems/DamageOnHighSpeedImpactSystem.cs:48]: Slower than before.

File: Content.Shared/Friction/TileFrictionController.cs [hunk 6]
```diff
@@ -102,21 +102,22 @@ public override void UpdateBeforeMapSolve(bool prediction, PhysicsMapComponent m
...
             float control;
 
+            var speed = velocity.Length();
             if (friction > 0.0f)
             {
                 // TBH I can't really tell if this makes a difference.
                 if (!prediction)
                 {
-                    control = speed < _stopSpeed ? _stopSpeed : speed;
+                    control = velocity.IsShorterThan(_stopSpeed) ? _stopSpeed : speed;
```
Comment (moonheart08) [inline Content.Shared/Friction/TileFrictionController.cs:120]: Slower than before.

File: Content.Shared/Movement/Systems/SharedMoverController.cs [hunk 7]
```diff
@@ -318,11 +318,11 @@ public void LerpRotation(EntityUid uid, InputMoverComponent mover, float frameTi
...
         private void Friction(float minimumFrictionSpeed, float frameTime, float friction, ref Vector2 velocity)
         {
-            var speed = velocity.Length();
-
-            if (speed < minimumFrictionSpeed)
+            if (velocity.IsShorterThan(minimumFrictionSpeed))
                 return;
 
+            var speed = velocity.Length();
+
```
Comment (moonheart08) [inline Content.Shared/Movement/Systems/SharedMoverController.cs:325]: Slower than before.

**Review by moonheart08 — CHANGES_REQUESTED** [review]
I think we'd be better off if you didn't change systems that did anything more complex with length than a comparison.

This is ultimately a micro-optimization that won't even show up in profiling, and you often mis-apply the optimization.

(This isn't meant to be mean, sorry. More of noting how actually useful this change is, which isn't very.)

File: Content.Client/Interaction/DragDropHelper.cs [hunk 8]
```diff
@@ -128,24 +128,24 @@ public void Update(float frameTime)
...
-                }
+                    if (!_onContinueDrag.Invoke(frameTime))
+                    {
+                        EndDrag();
+                    }
 
-                break;
-            }
+                    break;
+                }
```
[the remark this answers is not in the document]
Comment (Tayrtahn) [reply Content.Client/Interaction/DragDropHelper.cs:148]: Sorry, Visual Studio Code got worked up about it. I'll put it back.

**Comment by Tayrtahn** [comment]
> I think we'd be better off if you didn't change systems that did anything more complex with length than a comparison.
> 
> This is ultimately a micro-optimization that won't even show up in profiling, and you often mis-apply the optimization.
> 
> (This isn't meant to be mean, sorry. More of noting how actually useful this change is, which isn't very.)

No worries, I appreciate the feedback! It seems like I did indeed overestimate the cost of square roots - some quick research suggests that this change can have a significant impact in Unity (which is probably where I picked it up) but a pretty minimal effect outside of that specific environment.

I'll scale this back to just the most obvious use cases.

**Comment by moonheart08** [comment]
As I feel it'd be useful to you:
[Here is a distance comparison in sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgDkBXfGKASzFwBuGjWIBmJqQYBhBgG8aDJUwnAIEADYMAsgAoAajDAZoUhGgaHjphgE8AlPMXKXxAOwMELADIwAdgDmGAAWuo4AfHY+/kGh9sLULgC+zkriDGqaDGwGRiZQZhZW+VIOTokuyu6e0YEhAMoAjhzYsAAmYQyRtrWxTS3tYQnJNElAA=).

Let's assume the user has a ten year old Intel Broadwell microarchitecture processor, the instruction used for the sqrt operation is VSQRTPS and has a throughput of 1 every 14 clocks if uncontested. This is fine on it's own, pretty slow relative to LengthSquared (a single DPPS, or dot product operation, which can dispatch every other clock on the same processor)
In a hot loop, this optimization *would matter, possibly a lot*.

However, the estimate I gave is with the knowledge that most of these examples are in branch heavy, AVS-unit light code, which on an OoO/speculative processor like Broadwell (or newer architectures) means most of our time is going to be spent doing other things. So it's better to do the entire task once (compute length outright) or distribute the task (compute the dot product, then square root it later), but *never* do it more than once, as that opens room for actual pipeline contention instead of just waiting.

On newer processors (say, a modern Zen 4 one), the sqrt is even cheaper (10 clocks, none of which impede other instructions from running unless both ports 0 and 1 are occupied, which is guaranteed if we just computed the dot product of two vectors.), but we're still just juggling individual clocks here when the CPU is probably busy doing something more useful and working ahead while it computes the length.

**Comment by moonheart08** [comment]
I got the referenced instruction latencies from https://agner.org/optimize/, check the instruction table.

**Comment by moonheart08** [comment]
> > Measuring something being +-100 cycles faster/slower isn't possible.
> 
> Just repeat the benchmark enough times in series to exaggerate the difference. +- 100 cycles may be indistinguishable, but times a million nets something.

The difference immediately gets thrown out a window by OS process scheduling and CPU system action scheduling (modern processors can steal a core back from the OS to run system code in some circumstances)

**Comment by Tayrtahn** [comment]
The reasoning for it being slower makes sense to me. I'm going to revert the more complicated changes that delay the square root and just keep the ones that always remove.

**Comment by PJB3005** [comment]
The savings, while real, will not amount to anything compared to the logic that surrounds 99% of the changes in this PR. Yes, I know that without needing to set up a whole micro benchmark. It's for that reason that doing any complex modifications is not worth spending the energy on, and definitely not worth the risk of messing up and introducing a bug.

**Comment by Tayrtahn** [comment]
So I've changed it so it's basically* all one-liner changes using helper methods, which I think are at least as readable as the originals.

Are those worth keeping, or is this PR dead?

*there are also two loops that look for the shortest distance in a collection, but I can remove those too if they make the code harder to follow.

**Comment by metalgearsloth** [comment]
See above on "the optimisation is so small the risk of bugs outweighs any benefit", sorry.

**Pull request closed without merging**